entity-framework/core/internals/internal-service-provider.md

```diff
@@ -0,0 +1,166 @@
+---
+title: The internal service provider - EF Core
+author: ajcvickers
+ms.date: 01/25/2020
+uid: core/internals/internal-service-provider
+---
+
+# The internal service provider
+
+## Introduction
+
+Internally, EF Core is built around a [dependency injection]() container.
+This container exposes an extensible set of services that define the behavior of EF Core.
+Service implementations can come from the application, plug-ins, extensions, database providers, and EF Core itself.
+
+This container is exposed as an [`IServiceProvider`]() and implemented using [Microsoft.Extensions.DependencyInjection]().
+It is therefore known as the _internal service provider_.
+
+## DbContext and the internal service provider
+
+[`DbContext`]() instances are created by the application using:
+* Explicit construction with [`new`]()
+* The application's own dependency injection container
+This container is known as the [_application service provider_]().
+
+The DbContext instance then locates the internal service provider to use.<sup>[(1)](#footnote-1)</sup>
+The service provider is most commonly created automatically when needed.
+It is then [cached and reused]().
+
+An application can instead explicitly configure the internal service provider to use.
+The application must then handle caching.
+
+### Scopes and lifetimes
+
+A new [service provider scope]() is created for each DbContext instance.
+This means that:
+* [Singleton]() services are shared between DbContext instances
+* [Scoped]() services are shared within, but not between, DbContext instances
+
+A DbContext instance typically represents a [session]() or [unit-of-work]().
+This means scoped services work together to orchestrate the DbContext session.
+
+### Multiple internal service providers
+
+Most singleton services behave the same way between different context instances.
+That is, changes in the context configuration do not affect these services.
+
+However, some changes in context configuration require singleton services to behave differently.
+For example, [sensitive data logging]() requires singleton services to change the way they log.
+Likewise, some changes in context configuration require different service implementations.
+For example, [changing the database provider]() requires different implementations for many services.
+
+These additional service providers are cached automatically.
```

Comment:
Go into ISingletonOptions here or is it too much? Am worried this is a bit too abstract for newbies.

Reply:
I think that should go into a different page--I created a list of possible things to cover here: https://github.com/aspnet/EntityFramework.Docs/issues/1920